File: cpp/0083删除排序链表中的重复元素.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
};

ListNode *Create(vector<int> nums)
{
    if(!nums.size())
        return NULL;

    ListNode *head = new ListNode();
    ListNode *p = head;
    ListNode *q;
    p->val = nums[0];

    for (int i = 1;i<nums.size();++i)
    {
        q = new ListNode();
        q->val = nums[i];
        p->next = q;
        p = q;
    }
    p->next = NULL;
    return head;
}

void Show(ListNode *head)
{
    for (; head != NULL; head = head->next)
    {
        cout << head->val << " ";
    }
}

ListNode *deleteDuplicates(ListNode *head)
{
    if(!head)
        return head;
    ListNode *slow, *fast;
    slow = fast = head;

    while(fast)
    {
        if(slow->val!=fast->val)
        {
            slow->next = fast;
            slow = fast;
        }
        fast = fast->next;
    }
    slow->next = NULL;
    return head;
}

int main()
{
    vector<int> v = { 1,1,2};
    ListNode *head = Create(v);
    Show(deleteDuplicates(head));
}

File: python/0083删除排序链表中的重复元素.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def create(elems:list):
    if(not elems):
        return None
    head=ListNode()
    q=head
    q.val=elems[0]

    for i in elems[1:]:
        p=ListNode(i)
        q.next=p
        q=p
    return head

def show(head):
    while(head):
        print(head.val,end=" ")
        head=head.next

def deleteDuplicates(head:ListNode):
    if(not head):
        return head
    p=q=head
    
    while(q):
        if(p.val!=q.val):
            p.next=q
            p=q
        
        q=q.next
    p.next=None
    return head

head=create([12,3,4])
show(deleteDuplicates(head))